### PR TITLE
Fixing identity 4828

### PR DESCRIPTION
**Pull request by DMHP**
Implementing the jwks_uri

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    private static final String use = "sig";
+    private static final String kid = "d0ec514a32b6f88c0abd12a2840699bdd3deba9d";
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public String jwks(@PathParam("tenantDomain") String tenantDomain) {
+        RSAPublicKey publicKey = null;
+        JSONObject jwksJson = new JSONObject();
+        try {
+            if (StringUtils.isEmpty(tenantDomain)) {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: is it possible to use isBlank

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            jwksKeys.put("kid", kid);
+            jwksKeys.put("n", modulus);
+            jwksKeys.put("e", exponent);
+            jwksKeyArray.put(jwksKeys);
+            jwksJson.put("keys", jwksKeyArray);
+        } catch (Exception e) {
+            String errorMesage = "Error while generating the keyset";
+            log.error(errorMesage);
+            return errorMesage;
+        }
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: close FileInputStream in finally

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java (hunk 1 above)
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: Here it is not necessary to check isBlank as, if we get a white space as tenant domain it validates it as null. So StringUtils.isEmpty(tenantDomain) return true

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java [hunk 3]
```diff
@@ -99,6 +101,12 @@ public String jwks(@PathParam("tenantDomain") String tenantDomain) {
             String errorMesage = "Error while generating the keyset";
             log.error(errorMesage);
             return errorMesage;
+        }finally {
+            try {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: use IdentityIOStreamUtils in identity-core

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        RSAPublicKey publicKey = null;
+        JSONObject jwksJson = new JSONObject();
+        FileInputStream file = null;
+        try {
+            if (StringUtils.isEmpty(tenantDomain)) {
+
+                file = new FileInputStream(CarbonUtils.getServerConfiguration().getFirstProperty
+                        ("Security.KeyStore.Location"));
+                KeyStore keystore = KeyStore.getInstance(KeyStore.getDefaultType());
+                keystore.load(file, "wso2carbon".toCharArray());
```
Comment (chrisdrobison) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: I been following the progress on this feature lately for getting WSO2 IS OpenID Connect certified. Why is "wso2carbon" hard-coded? In our configuration, we don't use the wso2carbon name for the files, the certificate or the alias. Shouldn't all these be pull from config?

Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: Totally agreed. I think this has been already fixed in the master branch at [1] 

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java#L70

Comment (adamjbradley) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java]: Is there any way to access this without requiring Authentication?

**Pull request merged**